modules/phase_field/src/kernels/ACKappaFunction.C

```diff
@@ -0,0 +1,85 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ACKappaFunction.h"
+
+registerMooseObject("PhaseFieldApp", ACKappaFunction);
+
+template <>
+InputParameters
+validParams<ACKappaFunction>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Gradient energy term for when kappa as a function of the variable");
+  params.addParam<MaterialPropertyName>("mob_name", "L", "The mobility used with the kernel");
+  params.addParam<MaterialPropertyName>("kappa_name", "kappa_op", "The kappa function name");
+  params.addCoupledVar("v",
+                       "Vector of additional nonlinear variables that affect the gradient energy");
+  return params;
+}
+
+ACKappaFunction::ACKappaFunction(const InputParameters & parameters)
+  : DerivativeMaterialInterface<JvarMapKernelInterface<Kernel>>(parameters),
+    _var_name(getParam<NonlinearVariableName>("variable")),
+    _L_name(getParam<MaterialPropertyName>("mob_name")),
+    _L(getMaterialProperty<Real>(_L_name)),
+    _dLdvar(getMaterialPropertyDerivative<Real>(_L_name, _var_name)),
+    _kappa_name(getParam<MaterialPropertyName>("kappa_name")),
+    _dkappadvar(getMaterialPropertyDerivative<Real>(_kappa_name, _var_name)),
+    _d2kappadvar2(getMaterialPropertyDerivative<Real>(_kappa_name, _var_name, _var_name)),
+    _nv(coupledComponents("v")),
+    _v_name(_nv),
+    _grad_v(_nv),
+    _dLdv(_nv),
+    _d2kappadvardv(_nv)
+{
+  for (unsigned int i = 0; i < _nv; ++i)
+  {
+    _v_name[i] = getVar("v", i)->name();
+    _grad_v[i] = &coupledGradient("v", i);
+    _dLdv[i] = &getMaterialPropertyDerivative<Real>(_L_name, _v_name[i]);
+    _d2kappadvardv[i] = &getMaterialPropertyDerivative<Real>(_kappa_name, _var_name, _v_name[i]);
+  }
+}
+
+Real
+ACKappaFunction::computeQpResidual()
+{
+  return 0.5 * _test[_i][_qp] * _L[_qp] * _dkappadvar[_qp] * computeFg();
+}
+
+Real
+ACKappaFunction::computeQpJacobian()
+{
+  Real preJac = 0.5 * _test[_i][_qp] * _phi[_j][_qp] * computeFg();
+  Real term1 = _test[_i][_qp] * _L[_qp] * _dkappadvar[_qp] * _grad_u[_qp] * _grad_phi[_j][_qp];
+
+  return preJac * (_dLdvar[_qp] * _dkappadvar[_qp] + _L[_qp] * _d2kappadvar2[_qp]) + term1;
+}
+
+Real
+ACKappaFunction::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  const unsigned int i = mapJvarToCvar(jvar);
+  Real PreJac = 0.5 * _test[_i][_qp] * _phi[_j][_qp] * computeFg();
+  Real term1 =
+      _test[_i][_qp] * _L[_qp] * _dkappadvar[_qp] * (*_grad_v[i])[_qp] * _grad_phi[_j][_qp];
+  return PreJac * ((*_dLdv[i])[_qp] * _dkappadvar[_qp] + _L[_qp] * (*_d2kappadvardv[i])[_qp]) +
+         term1;
+}
+
+Real ACKappaFunction::computeFg() // gradient energy term
+{
+  Real sum_grad_etai2 = 0.0;
+  for (unsigned int i = 0; i < _nv; ++i)
+    sum_grad_etai2 += (*_grad_v[i])[_qp] * (*_grad_v[i])[_qp];
+
+  Real grad_var2 = _grad_u[_qp] * _grad_u[_qp];
+  return sum_grad_etai2 + grad_var2;
+}
```

Comment:
Wait!! Why is this sum over all the gradient variables? In general, kappa values corresponding to different coupled variables can be different, right?

Reply:
The idea with the kernel is that you are using a functional kappa value instead of different kappas for each variable. It achieves a similar effect to using multiple kappas, but with transition regions. I can't think of a scenario where someone would want multiple functional kappas, but if they did they could try being selective on what variables to include in "v".

Reply:
Here is what I am trying to understand, the gradient energy term is 
![grad1](https://user-images.githubusercontent.com/9493686/43477905-a64723fc-94b9-11e8-9f00-d5220cccdb37.gif)
and you are implementing its derivative w.r.t the order parameters. So, shouldn't the _generic_ implementation be 
![codecogseqn 2](https://user-images.githubusercontent.com/9493686/43479850-b18511d4-94be-11e8-9086-00d6bd4b7083.gif)
instead of 
![codecogseqn 1](https://user-images.githubusercontent.com/9493686/43479837-a8747468-94be-11e8-8180-487536848d48.gif) ? 
It assumes same kappa function for all the coupled order parameters, cases such as polycrystal solidification can have different kappa functions.

Reply:
That would be the most generic case, but
1. it would be a huge headache for me to rewrite the kernel
2. the UI for the kernel would be more complicated and the kernel harder to use
3. the feature would (almost) never be used,
4. in the rare case that it is needed, you can achieve the same effect by using the kernel multiple times with a single coupled eta each time, effectively doing the sum yourself.

I looked up a couple of polycrystal solidification papers and could not find one with multiple functional kappas. [One](https://www.sciencedirect.com/science/article/pii/S1359645403003884) used multiple constant kappas, and [another](https://www.sciencedirect.com/science/article/pii/S1359645409008350) used a single functional kappa. Can you send a link to one with multiple functional kappas? I'm curious about the scenario where that would be desirable.

Reply:
I understand. I am working on something similar to what you have found and intended to implement some associated kernels (#11838). It's okay, I can work around it for now. I was curious to know why the kernel is not implemented in the generic form. Thanks for the clarification.

Reply:
This kernel could be expanded in a later PR (even by you @SudiptaBiswas)

Reply:
Yeah, I would work on it later.